app/Http/Controllers/SettingController.php

```diff
@@ -2,9 +2,86 @@
 
 namespace App\Http\Controllers;
 
+use App\Http\Requests\StoreSettingRequest;
+use App\Models\Setting;
+use App\Models\User;
 use Illuminate\Http\Request;
+use Illuminate\Validation\Rule;
 
 class SettingController extends Controller
 {
-    //
+    public function index()
+    {
+        // get Sam to consolidate this, it could be improved, maybe a different way instead of get()?
+        return view('settings.index',[
+            'settings' => auth()->user()->settings()->get()
+        ]);
+
+    }
+
+    public static function create(User $user, string $key, string $value)
+    {
+        //creates a user setting from the parameters. use in code only. will be redundant probably.
+
+        $user->settings()->create([
+            'user_id' => $user->id,
+            'key' => $key,
+            'value' => $value
+        ]);
+
+    }
+
+    public function store(StoreSettingRequest $request)
+    {
+        //creates a user setting for the currently logged in user /w validation
+
+
+        //TODO: complete authorization in StoreSettingRequest
+        $validated = $request->validated();
+
+        auth()->user()->settings()->create(
+            $validated
+        );
+
+        return redirect('settings')->with('success', 'Setting created :)');
+    }
+
+    public function read(string $key)
+    {
+        //TODO: fix else return/throw exception
+        $settings = auth()->user()->settings()->get()->whereStrict('key', $key);
```

Comment:
Oh wow, I did not know about `whereStrict`! That's cool. 

One improvement you can make to this would be to put the "where" before the `get()` because you are currently.

1. Grabbing all of the results
2. Filtering the collection by the key

But you can improve it by

1. Getting all the results where the key is "X"
2. Returning the collection

```php
$settings = auth()->user()->settings()->where('key', $key)->get();
```

Also because you only have one key, you can use `first`, `firstOrFail` or `sole` instead.

- first will return the model or NULL
- first will return the model or throw a ModelNotFoundException which appears as a 404 if rendered to the user
- sole will return the model or throw a ModelNotFoundException or will throw a TooManyFound exception if more than 1 record exists for that query.

I think this would be correct:

```php
$setting = auth()->user()->settings()->where('key', $key)->firstOrFail();
```

Reply:
Alternatively, you can actually use route model binding to remove the requirement to query completely. Let's talk about this tomorrow.